fabric-chaincode-shim/src/main/java/org/hyperledger/fabric/shim/ChatChaincodeWithPeer.java

```diff
@@ -0,0 +1,50 @@
+/*
+ * Copyright 2020 IBM All Rights Reserved.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.fabric.shim;
+
+import io.grpc.stub.StreamObserver;
+import org.hyperledger.fabric.Logging;
+import org.hyperledger.fabric.protos.peer.ChaincodeGrpc;
+import org.hyperledger.fabric.protos.peer.ChaincodeShim;
+
+import java.io.IOException;
+import java.util.logging.Logger;
+
+public class ChatChaincodeWithPeer extends ChaincodeGrpc.ChaincodeImplBase {
+    private static Logger logger = Logger.getLogger(ChatChaincodeWithPeer.class.getName());
+
+    private ChaincodeBase chaincodeBase;
+
+    ChatChaincodeWithPeer(final ChaincodeBase chaincodeBase) throws IOException {
+        if (chaincodeBase == null) {
+            throw new IOException("chaincodeBase can't be null");
+        }
+        chaincodeBase.validateOptions();
+
+        this.chaincodeBase = chaincodeBase;
```

Comment:
In the ContractRouter.main() method there are some other calls that need to made on the ContractRouter before it's able to work.  This looks like the correct place to put the,.

Reply:
I agree that it's a good place
but I see also two issues there - those methods called in ContractRouter.main() are private and protected - not available there and I'm not sure if extending access modifiers for them is a good idea
2) here we work with ChaincodeBase abstract class and if in the future there will be another implementation of it (not only ContractRouter) - there're possible troubles IMO

what if just create one more public method in ContractRouter which will contain calls from .main() and will allow chaincode developers to start routing manually in chaincode together with starting server?

WDYT, @mbwhite?

Reply:
Yes; I agree with the solution you've suggested - what I would do myself. 

There won't be another subclass; the Contract Router is the preferred way forward.